File: ios/AccountBridge.m
#import "AccountBridge.h"

#if __has_include(<React/RCTBridge.h>)
#import <React/RCTBridge.h>
#elif __has_include(“RCTBridge.h”)
#import “RCTBridge.h”
#else
#import “React/RCTBridge.h”
#endif

#import <Textile/TextileApi.h>
#import "utils.h"

#define SYSTEM_VERSION_LESS_THAN(v) ([[[UIDevice currentDevice] systemVersion] compare:v options:NSNumericSearch] == NSOrderedAscending)

@interface AccountBridge ()

@property (atomic, strong) MobileSearchHandle *searchHandle;

@end

@implementation AccountBridge

RCT_EXPORT_MODULE();

RCT_EXPORT_METHOD(address:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(Textile.instance.account.address);
}

RCT_EXPORT_METHOD(seed:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  resolve(Textile.instance.account.seed);
}

RCT_EXPORT_METHOD(encrypt:(NSString*)inputStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSData *data = [Textile.instance.account encrypt:[[NSData alloc] initWithBase64EncodedString:inputStr options:0] error:&error];
  fulfillWithResultAndNilDefault([data base64EncodedStringWithOptions:0], @"", error, resolve, reject);
}

RCT_EXPORT_METHOD(decrypt:(NSString*)inputStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  NSData *data = [Textile.instance.account decrypt:[[NSData alloc]initWithBase64EncodedString:inputStr options:0] error:&error];
  fulfillWithResultAndNilDefault([data base64EncodedStringWithOptions:0], @"", error, resolve, reject);
}

RCT_EXPORT_METHOD(contact:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  NSError *error;
  Contact *contact = [Textile.instance.account contact:&error];
  fulfillWithResultAndNilDefault([contact.data base64EncodedStringWithOptions:0], @"", error, resolve, reject);
}

RCT_EXPORT_METHOD(sync:(NSString*)optionsStr resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  if (self.searchHandle) {
    [self.searchHandle cancel];
    self.searchHandle = nil;
  }
  NSError *error;
  NSData *optionsData = [[NSData alloc] initWithBase64EncodedString:optionsStr options:0];
  QueryOptions *options = [[QueryOptions alloc] initWithData:optionsData error:&error];
  self.searchHandle = [Textile.instance.account sync:options error:&error];
  fulfillWithResult(nil, error, resolve, reject);
}

RCT_EXPORT_METHOD(cancelSync:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject) {
  if (self.searchHandle) {
    [self.searchHandle cancel];
    self.searchHandle = nil;
  }
  resolve(@"success");
}

@end

@implementation RCTBridge (AccountBridge)

- (AccountBridge *)accountBridge {
  return [self moduleForClass:[AccountBridge class]];
}

@end

File: android/src/main/java/io/textile/rnmobile/AccountBridge.java
package io.textile.rnmobile;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import io.textile.pb.Model;
import io.textile.pb.QueryOuterClass;
import io.textile.textile.Textile;
import mobile.SearchHandle;

public class AccountBridge extends ReactContextBaseJavaModule {

    private static SearchHandle searchHandle;

    private Executor executor = Executors.newSingleThreadExecutor();

    public AccountBridge(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "AccountBridge";
    }

    @ReactMethod
    public void address(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.instance().account.address());
                }
                catch (Exception e) {
                    promise.reject("address", e);
                }
            }
        });
    }

    @ReactMethod
    public void seed(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Textile.instance().account.seed());
                }
                catch (Exception e) {
                    promise.reject("seed", e);
                }
            }
        });
    }

    @ReactMethod
    public void encrypt(final String inputStr, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Util.encode(Textile.instance().account.encrypt(Util.decode(inputStr))));
                }
                catch (Exception e) {
                    promise.reject("encrypt", e);
                }
            }
        });
    }

    @ReactMethod
    public void decrypt(final String inputStr, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    promise.resolve(Util.encode(Textile.instance().account.decrypt(Util.decode(inputStr))));
                }
                catch (Exception e) {
                    promise.reject("decrypt", e);
                }
            }
        });
    }

    @ReactMethod
    public void contact(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    Model.Contact contact = Textile.instance().account.contact();
                    promise.resolve(Util.encode(contact.toByteArray()));
                }
                catch (Exception e) {
                    promise.reject("contact", e);
                }
            }
        });
    }

    @ReactMethod
    public void sync(final String optionsStr, final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    if (AccountBridge.searchHandle != null) {
                        AccountBridge.searchHandle.cancel();
                    }
                    QueryOuterClass.QueryOptions options = QueryOuterClass.QueryOptions.parseFrom(Util.decode(optionsStr));
                    AccountBridge.searchHandle = Textile.instance().account.sync(options);
                    promise.resolve(AccountBridge.searchHandle.getId());
                }
                catch (Exception e) {
                    promise.reject("syncAccount", e);
                }
            }
        });
    }

    @ReactMethod
    public void cancelSync(final Promise promise) {
        executor.execute(new Runnable() {
            @Override
            public void run() {
                try {
                    if (AccountBridge.searchHandle != null) {
                        AccountBridge.searchHandle.cancel();
                        AccountBridge.searchHandle = null;
                    }
                    promise.resolve(null);
                }
                catch (Exception e) {
                    promise.reject("cancelSync", e);
                }
            }
        });
    }
}
